plugins/github-release-manager/src/api/PluginApiClient.ts

```diff
@@ -0,0 +1,961 @@
+    bumpedTag: string;
+    selectedPatchCommit: UnboxArray<
+      UnboxReturnedPromise<IPluginApiClient['getRecentCommits']>
+    >;
+    mergeTree: string;
+    releaseBranchSha: string;
+  } & OwnerRepo,
+) => Promise<{
+  message: string;
+  sha: string;
+}>;
+export type CreateCherryPickCommitResult = UnboxReturnedPromise<CreateCherryPickCommit>;
+
+type ReplaceTempCommit = (
+  args: {
+    releaseBranchName: string;
+    cherryPickCommit: UnboxReturnedPromise<
+      IPluginApiClient['patch']['createCherryPickCommit']
+    >;
+  } & OwnerRepo,
+) => Promise<{
+  ref: string;
+  object: {
+    sha: string;
+  };
+}>;
+export type ReplaceTempCommitResult = UnboxReturnedPromise<ReplaceTempCommit>;
+
+type CreateTagObject = ({
+  bumpedTag,
+  updatedReference,
+}: {
+  bumpedTag: string;
+  updatedReference: ReplaceTempCommitResult;
+} & OwnerRepo) => Promise<{
+  tag: string;
+  sha: string;
+}>;
+export type CreateTagObjectResult = UnboxReturnedPromise<CreateTagObject>;
+
+type CreateReference = (
+  args: {
+    bumpedTag: string;
+    createdTagObject: CreateTagObjectResult;
+  } & OwnerRepo,
+) => Promise<{
+  ref: string;
+}>;
+export type CreateReferenceResult = UnboxReturnedPromise<CreateReference>;
+
+type UpdateRelease = (
+  args: {
+    bumpedTag: string;
+    latestRelease: NonNullable<GetLatestReleaseResult>;
+    tagParts: SemverTagParts | CalverTagParts;
+    selectedPatchCommit: GetRecentCommitsResultSingle;
+  } & OwnerRepo,
+) => Promise<{
+  name: string | null;
+  tagName: string;
+  htmlUrl: string;
+}>;
+export type UpdateReleaseResult = UnboxReturnedPromise<UpdateRelease>;
+
+/**
+ * PromoteRc
+ */
+type PromoteRelease = (
+  args: {
+    releaseId: NonNullable<GetLatestReleaseResult>['id'];
+    releaseVersion: string;
+  } & OwnerRepo,
+) => Promise<{
+  name: string | null;
+  tagName: string;
+  htmlUrl: string;
+}>;
+export type PromoteReleaseResult = UnboxReturnedPromise<PromoteRelease>;
+
+/**
+ * Stats
+ */
+type GetAllTags = (
+  args: OwnerRepo,
+) => Promise<
+  Array<{
+    tagName: string;
+    sha: string;
+  }>
+>;
+export type GetAllTagsResult = UnboxReturnedPromise<GetAllTags>;
+
+type GetAllReleases = (
+  args: OwnerRepo,
+) => Promise<
+  Array<{
+    id: number;
+    name: string | null;
+    tagName: string;
+    createdAt: string | null;
+    htmlUrl: string;
+  }>
+>;
+export type GetAllReleasesResult = UnboxReturnedPromise<GetAllReleases>;
+
+type GetCommit = (
+  args: {
+    ref: string;
+  } & OwnerRepo,
+) => Promise<{
+  createdAt: string | undefined;
+}>;
+export type GetCommitResult = UnboxReturnedPromise<GetCommit>;
+
+export interface IPluginApiClient {
+  getHost: GetHost;
+  getRepoPath: GetRepoPath;
+  getOwners: GetOwners;
+  getRepositories: GetRepositories;
+  getUsername: GetUsername;
+  getRecentCommits: GetRecentCommits;
+  getLatestRelease: GetLatestRelease;
+  getRepository: GetRepository;
+  getLatestCommit: GetLatestCommit;
+  getBranch: GetBranch;
+  createRc: {
```

Comment:
Would probably spell this out tbh, as in `createReleaseCandidate`. Also not sure about the nested objects in the API here, would typically prefer a flat API. Continuing on the discussion above about the level of this API, maybe all of these should be merged into a single `createReleaseCandidate` method?

Reply:
> Would probably spell this out tbh, as in createReleaseCandidate

Sure 👍 

> Also not sure about the nested objects in the API here, would typically prefer a flat API

Yea, I've considered this as well. Before I do that I'd like to generalize some of the API methods first so that they're more generic.

> maybe all of these should be merged into a single createReleaseCandidate method?

Will respond to this in https://github.com/backstage/backstage/pull/5299#discussion_r622333825

Reply:
> Also not sure about the nested objects in the API here, would typically prefer a flat API

Will take a stab at flattening the API :sli